File: 167 Two Sum II - Input Array Is Sorted/167.two-sum-ii-input-array-is-sorted.py
#
# @lc app=leetcode id=167 lang=python3
#
# [167] Two Sum II - Input Array Is Sorted
#

# @lc code=start
class Solution:
    def twoSum(self, numbers: List[int], target: int) -> List[int]:
        # 1. Initialize two pointers: left starting at the beginning (0) and right at the end (length - 1)
        left, right = 0, len(numbers) - 1
        # 2. Iterate while left pointer is less than right pointer
        while left < right:
            # 3. Calculate the sum of the numbers at the left and right pointers
            current_sum = numbers[left] + numbers[right]
            # 4. Check if the current sum is equal to the target
            if current_sum == target:
                # 5. If equal, return the indices adjusted for 1-indexing
                return [left + 1, right + 1]
            # 6. If the current sum is less than the target, move the left pointer to the right
            elif current_sum < target:
                left += 1
            # 7. If the current sum is greater than the target, move the right pointer to the left
            else:
                right -= 1
# @lc code=end

File: 167 Two Sum II - Input Array Is Sorted/167.two-sum-ii-input-array-is-sorted.cpp
/*
 * @lc app=leetcode id=167 lang=cpp
 *
 * [167] Two Sum II - Input Array Is Sorted
 */

// @lc code=start
#include <vector>
#include <iostream>

class Solution {
public:
    vector<int> twoSum(vector<int>& numbers, int target) {
        int left = 0;  // 1. Initialize the left pointer at the start of the array
        int right = numbers.size() - 1;  // 2. Initialize the right pointer at the end of the array

        while (left < right) {  // 3. Continue until the pointers meet
            int current_sum = numbers[left] + numbers[right];  // 4. Calculate the sum of the two pointed numbers

            if (current_sum == target) {  // 5. Check if the current sum is equal to the target
                return {left + 1, right + 1};  // 6. Return indices (1-indexed)
            } 
            else if (current_sum < target) {  // 7. If the sum is less than the target, move the left pointer right
                left++;
            } 
            else {  // 8. If the sum is greater than the target, move the right pointer left
                right--;
            }
        }
        return {};  // This line should never be reached as per problem statement.
    }
};
// @lc code=end

